python/ml4ir/applications/ranking/data/scripts/create_dataset.py

```diff
@@ -0,0 +1,227 @@
+
+    # Get names of essential features from feature_config
+    feature_query_key = feature_config.get_query_key('name')
+    feature_rank = feature_config.get_rank('name')
+    feature_label = feature_config.get_label('name')
+
+    # For ranking data, calculate number of results from data and filter accordingly
+    seed_data, nres = filter_nres(seed_data, feature_query_key, feature_rank, feature_nres, max_num_records, logger)
+
+    # For ranking data, calculate clicked rank from seed data as label_rank
+    seed_data = pd.merge(seed_data,
+                         seed_data[seed_data[feature_label]==True][[feature_rank,feature_query_key]].rename(columns={
+                             feature_rank:'label_rank'}),
+                         on=feature_query_key,
+                         how='left')
+
+    # Now generate data distributions from sample data
+    for mycol in list(seed_data.columns):
+        seed_dict[mycol] = list(seed_data[mycol].values)
+        seed_dict_labeled_pos[mycol] = list(seed_data[seed_data[feature_label] == 1][mycol].values)
+        seed_dict_labeled_neg[mycol] = list(seed_data[seed_data[feature_label] == 0][mycol].values)
+
+    for _ in range(num_samples):
+        # Generate a synthetic query key
+        q_id = generate_key(seed_dict, feature_config.get_query_key('name'))
+
+        # Sample to choose the number of members of the query key (in ranking, the number of results)
+        q_nseq = random.sample(seed_dict[nres], 1)[0]
+
+        # Sample to choose which member is labeled positive (in ranking, which result is clicked),
+        # sampling only from query keys with the same nseq (number of results)
+        q_labelrank = random.sample(list(seed_data[seed_data[nres]==q_nseq].label_rank.values), 1)[0]
+
+        # Build characteristics for each member of the query key (in ranking, each result of the query)
+        for i in range(int(q_nseq)):
+            is_labeled_pos = (i + 1 == q_labelrank)
+            row_dict = {
+                feature_query_key: q_id,
+                nres: q_nseq,
+                'label_rank': q_labelrank,
+                feature_rank: i + 1
+            }
+            for mycol in list(seed_data.columns):
+                # Could also let user specify distribution to sample from for each feature
+                if mycol not in row_dict.keys() and is_labeled_pos:
+                    # row_dict[mycol] = random.sample(seed_dict_labeled_pos[mycol])
+                    row_dict[mycol] = None  # Default value, check type # Could also be functions we pass in
+                elif mycol not in row_dict.keys() and not is_labeled_pos:
+                    # row_dict[mycol] = random.sample(seed_dict_labeled_neg[mycol])
+                    row_dict[mycol] = None
+
+            row_dict = add_feature_highval(row_dict, feature_highval, is_labeled_pos, q_labelrank)
+
+            rows_df.append(row_dict)
+    return pd.DataFrame(rows_df)
+
+
+def main():
```

Comment:
@calliebradley Can you add a comment note here on how to run this? And maybe an example as well?

Reply:
@lastmansleeping Thanks, there's a comment at the top actually, noting how to run, both here and in test_create_dataset.py. For running from here with example data, you'll need to download the example data and feature_config.yaml that I noted in the Gus story, they are posted here: /projects/search_relevance/ml4ir_support_files/synth_data

You don't need to download anything extra to run the test_create_dataset.py though :)